### PR TITLE
feat: send last textDocument/diagnostic id

### PR DESCRIPTION
**Pull request by drzbida**
Related to: https://github.com/seblj/roslyn.nvim/issues/93

Tested:
NVIM v0.10.2
NVIM v0.11.0-dev-1287+g8f84167c30
Custom on attach in the configuration
Custom handler for "textDocument/diagnostic" in the configuration

Unfortunately it seems 0.10 and 0.11 need slightly different implementations. Although from what I understand, 0.11 version should work in 0.10,  it doesn't. It receives a table instead of the method name and other parameters are received wrong too. 

Note: I'm not sure what are the situations when the received handler from the request method is not nil, it looks like even with a custom handler for textDocument/diagnostic it receives nil.

### CODE REVIEW

File: lua/roslyn/lsp.lua [hunk 1]
```diff
@@ -82,6 +86,50 @@ function M.start(bufnr, root_dir, on_init)
         end
     end
 
+    config.on_attach = function(client, attach_bufnr)
+        local original_request = client.request
+        local last_diagnostic = nil
+        local default_handler = roslyn_config.config.handlers["textDocument/diagnostic"]
```
Comment (seblj) [inline lua/roslyn/lsp.lua:92]: `roslyn.config.handlers` might be nil if the user doesn't send in any `handlers` in the setup. We should just set `handlers` to an empty table in the default config in `config.lua`

File: lua/roslyn/lsp.lua [hunk 2]
```diff
@@ -82,6 +86,50 @@ function M.start(bufnr, root_dir, on_init)
...
+            or vim.lsp.handlers["textDocument/diagnostic"]
+
+        if vim.fn.has("nvim-0.11") == 1 then
```
Comment (seblj) [inline lua/roslyn/lsp.lua:95]: Doesn't this also work with nvim-0.10?😔 I was under the impression that it should work when I looked through the PR for deprecation of `client.request` in favor of `client:request`... Anywas, it is fine if it doesn't work, but it's just sad...

**Review by seblj — COMMENTED** [review]
It looks good👍 Should at least fix the comment about handlers being potentially being `nil`

**Comment by seblj** [comment]
Btw, do you know about the identifier which was also used by vscode? No need to include it in this PR though if you don't know of the top of your head and isn't as easy fix

File: lua/roslyn/lsp.lua (hunk 2 above)
Comment (seblj) [reply lua/roslyn/lsp.lua:95]: Okay yeah I just re-read the comments in the PR now...

**Comment by drzbida** [comment]
> Btw, do you know about the identifier which was also used by vscode? No need to include it in this PR though if you don't know of the top of your head and isn't as easy fix

I've looked a few days ago at the implementation from roslyn (not in too much detail) and bookmarked a few things. Here are the relevant parts

https://github.com/dotnet/roslyn/blob/517ce1a9ff86aa1b54c4bcb604d2753d05492fb2/src/LanguageServer/Protocol/Handler/Diagnostics/Public/PublicDocumentPullDiagnosticsHandler.cs#L20

The identifier is used for a [category](https://github.com/dotnet/roslyn/blob/main/src/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs#L122)

```csharp
protected override string? GetRequestDiagnosticCategory(DocumentDiagnosticParams diagnosticsParams)
        => diagnosticsParams.Identifier;
```

[And the category is used for this cache](https://github.com/dotnet/roslyn/blob/main/src/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs#L53)

```csharp
/// <summary>
    /// Map of diagnostic category to the diagnostics cache for that category.
    /// Each category has a separate cache as they have disjoint resultIds and diagnostics.  For example, we may have
    /// one cache for DocumentSyntax, another for DocumentSemantic, another for WorkspaceSemantic, etc etc.
    /// </summary>
    private readonly ConcurrentDictionary<string, DiagnosticsPullCache> _categoryToVersionedCache = [];
```

It is used here for diagnostics
```csharp
var handlerName = $"{this.GetType().Name}(category: {category})";
context.TraceInformation($"{handlerName} started getting diagnostics");

var versionedCache = _categoryToVersionedCache.GetOrAdd(handlerName, static handlerName => new(handlerName));
```

I'm not really sure, but at first glance I don't think this one has any performance impacts for us.

**Comment by seblj** [comment]
I'm thinking that we can wait with this until it is implemented in neovim core, as cancelling the pending requests seems to be the biggest culprit, and this didn't seem to have so much effect. What do you think?

**Pull request closed without merging**